substrate/frame/staking/ah-client/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+		) -> DispatchResult {
+			// TODO: check origin
+
+			// Save the validator set. We don't care if there is a validator set which was not used.
+			ValidatorSet::<T>::put(Some(new_validator_set));
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> historical::SessionManager<T::AccountId, Exposure<T::AccountId, BalanceOf<T>>>
+		for Pallet<T>
+	{
+		fn new_session(
+			new_index: sp_staking::SessionIndex,
+		) -> Option<Vec<(T::AccountId, Exposure<T::AccountId, BalanceOf<T>>)>> {
+			// todo: check if we need to keep a copy of the validator set in case of another call to
+			// `new_session` before we get new validators. My assumption right now is that
+			// returning `None` will cause validator set to remain unchanged.
+			ValidatorSet::<T>::take()
+		}
+
+		// This method is supposed to be used by
+		fn new_session_genesis(
+			new_index: SessionIndex,
+		) -> Option<Vec<(T::AccountId, Exposure<T::AccountId, BalanceOf<T>>)>> {
+			ValidatorSet::<T>::take()
+		}
+
+		fn start_session(start_index: SessionIndex) {
+			<Self as pallet_session::SessionManager<_>>::start_session(start_index)
+		}
+
+		fn end_session(end_index: SessionIndex) {
+			<Self as pallet_session::SessionManager<_>>::end_session(end_index)
+		}
+	}
+
+	impl<T: Config> pallet_session::SessionManager<T::AccountId> for Pallet<T> {
+		fn new_session(_: u32) -> Option<Vec<<T as frame_system::Config>::AccountId>> {
+			// Doesn't do anything because all the logic is handled in `historical::SessionManager`
+			// implementation
+			defensive!("new_session should not be called");
+			None
+		}
+
+		fn end_session(session_index: u32) {
+			// todo: pass era points info
+
+			let message = Xcm(vec![
+				Instruction::UnpaidExecution {
+					weight_limit: WeightLimit::Unlimited,
+					check_origin: None,
+				},
+				mk_asset_hub_call::<T>(StakingCalls::EndSession(session_index)),
+			]);
+			if let Err(err) = send_xcm::<T::SendXcm>(
+				Location::new(0, [Junction::Parachain(T::AssetHubId::get())]),
+				message,
+			) {
+				log::error!(target: LOG_TARGET, "Sending `EndSession` to AssetHub failed: {:?}", err);
+			}
+		}
+
+		fn start_session(session_index: u32) {
+			// todo: pass active validator set somehow(tm)
+
+			let message = Xcm(vec![
+				Instruction::UnpaidExecution {
+					weight_limit: WeightLimit::Unlimited,
+					check_origin: None,
+				},
+				mk_asset_hub_call::<T>(StakingCalls::StartSession(session_index)),
+			]);
+			if let Err(err) = send_xcm::<T::SendXcm>(
+				Location::new(0, [Junction::Parachain(T::AssetHubId::get())]),
+				message,
+			) {
+				log::error!(target: LOG_TARGET, "Sending `StartSession` to AssetHub failed: {:?}", err);
+			}
+		}
+	}
+
+	impl<T> pallet_authorship::EventHandler<T::AccountId, BlockNumberFor<T>> for Pallet<T>
+	where
+		T: Config + pallet_authorship::Config + pallet_session::Config,
+	{
+		fn note_author(author: T::AccountId) {
+			// save the account id in a storage item
+			todo!()
```

Comment:
@kianenigma one question here.

I want to accumulate all authors in a storage item and then send them to AH with a XCM message. 

`AccountId` is set in the configuration of both pallets on both networks. As far as I know the message definition ([this](https://github.com/paritytech/polkadot-sdk/pull/7357/files#diff-dc472b745eb66637ab425971f616728efb2fac075758a72a91fa797f262a57b2R51)) should be something concrete (correct?).  Is there an elegant way to handle this without hardcoding a type in the XCM message and hoping `AccountId` on AH and RC to be configured correctly?